src/providers/swift.rs

```diff
@@ -0,0 +1,147 @@
+use super::Provider;
+use crate::nixpacks::{
+    app::App,
+    environment::Environment,
+    nix::pkg::Pkg,
+    phase::{BuildPhase, InstallPhase, SetupPhase, StartPhase},
+};
+use anyhow::Result;
+use indoc::formatdoc;
+
+static LATEST_SWIFT: &str = "5.6.1";
+
+pub struct SwiftProvider {}
+
+impl Provider for SwiftProvider {
+    fn name(&self) -> &str {
+        "swift"
+    }
+
+    fn detect(&self, app: &App, _env: &Environment) -> Result<bool> {
+        Ok(app.includes_file("Package.swift")
+            || (app.includes_file("Package.swift") && app.includes_file("Package.resolved")))
+    }
+
+    fn setup(&self, _app: &App, _env: &Environment) -> Result<Option<SetupPhase>> {
+        let pkgs = vec![Pkg::new("python38"), Pkg::new("wget")];
+
+        Ok(Some(SetupPhase::new(pkgs)))
+    }
+
+    fn install(&self, app: &App, _env: &Environment) -> Result<Option<InstallPhase>> {
+        let version = SwiftProvider::get_swift_version(&app)?;
+        let (download_url, name) = SwiftProvider::make_download_url(version);
+
+        // https://forums.swift.org/t/which-clang-package-should-we-install/20542/14
+        let install_cmd = formatdoc! {"
+        sudo apt-get update && \
+        sudo apt-get install -y build-essential clang libsqlite3-0 libncurses6 libcurl4 libxml2 libatomic1 libedit2 libsqlite3-0 libcurl4 libxml2 libbsd0 libc6-dev && \
+        wget -q {download_url} && \
+        tar -xf {name}.tar.gz && \
+        sudo mv {name} /usr/share/swift
+        ",
+        name=name,
+        download_url=download_url
+        };
+
+        let mut install_phase = InstallPhase::new(install_cmd);
+
+        install_phase.add_path("/usr/share/swift/usr/bin".to_string());
+
+        Ok(Some(install_phase))
+    }
+
+    fn build(&self, _app: &App, _env: &Environment) -> Result<Option<BuildPhase>> {
+        Ok(Some(BuildPhase::new(
+            "swift build -c release --static-swift-stdlib".to_string(),
+        )))
+    }
+
+    fn start(&self, app: &App, _env: &Environment) -> Result<Option<StartPhase>> {
+        let raw_paths = app.find_files("Sources/**/main.swift")?;
+        let paths = raw_paths
+            .iter()
+            .filter(|&path| !path.to_string_lossy().contains(".build"))
```

Comment:
Can't we just check if the path name is .build?

Reply:
Can't actually, e.g. you're in the `/workspaces/nixpacks` directory, `find_files` returns the full path of the files that match the glob and dependencies also have `main.swift` files e.g. `/workspaces/nixpacks/examples/swift-vapor/.build/checkouts/swift-collections/Benchmarks/benchmark-tool/main.swift`

Reply:
Use .has_match to check if the folder exists. I think this should work for directories.

Reply:
How? IIRC you can't use exclude patterns on glob and what's exactly wrong with iterating over the paths?

Reply:
In the run code later down the line, you just do ./build so I thought that logic could be applied here.

Reply:
Oh, that logic cant be applied sadly, since the executable target is configured via the `Package.swift` file and I dont how to parse that.

Reply:
Could a regex be used?

Reply:
> Could a regex be used?

Possibly, but I'm not really well-versed in regex.

Reply:
Also the field varies on different version so it'll be quite hard i think

Reply:
That's fine. I can potentially give you a regex.

Reply:
If I may ask; what's wrong with my current method of finding the executable name? Since there can be only one `main.swift` file anyway.

Reply:
From my knowledge, any file can be the mainfile. This is why I thought of making it dynamic so it would support more cases, if that makes sense.

Reply:
sure then up to you lol

Reply:
resolved in a discord thread